### ISSUE TITLE
Regresion Lineal

### ISSUE THREAD

**Issue by HeladeroTV** [HeladeroTV/Tareas-IA#4]
``import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Datos de ejemplo: horas de estudio vs calificación
# X: horas de estudio (variable independiente)
# y: calificación obtenida (variable dependiente)
X = np.array([1, 2, 3, 4, 5, 6, 7, 8]).reshape(-1, 1)
y = np.array([65, 68, 72, 75, 78, 82, 86, 89])

# Crear y entrenar el modelo de regresión lineal
modelo = LinearRegression()
modelo.fit(X, y)

# Realizar predicciones con el modelo entrenado
y_pred = modelo.predict(X)

# Evaluar el modelo
mse = mean_squared_error(y, y_pred)
r2 = r2_score(y, y_pred)

# Imprimir coeficientes y evaluación
print(f"Intercepto (b₀): {modelo.intercept_:.2f}")
print(f"Pendiente (b₁): {modelo.coef_[0]:.2f}")
print(f"Error cuadrático medio: {mse:.2f}")
print(f"Coeficiente de determinación (R²): {r2:.2f}")

# Visualizar los resultados
plt.figure(figsize=(10, 6))
plt.scatter(X, y, color='blue', label='Datos reales')
plt.plot(X, y_pred, color='red', linewidth=2, label='Línea de regresión')
plt.title('Regresión Lineal: Horas de Estudio vs Calificación')
plt.xlabel('Horas de estudio')
plt.ylabel('Calificación')
plt.legend()
plt.grid(True)
plt.savefig('regresion_lineal.png')
plt.show()

# Ejemplo de predicción para un nuevo valor
horas_nuevas = np.array([[9]])
prediccion = modelo.predict(horas_nuevas)
print(f"Predicción para {horas_nuevas[0][0]} horas de estudio: {prediccion[0]:.1f} puntos")``

Regresion Lineal Teorico:

Ejemplo teórico de regresión lineal

Tenemos 8 estudiantes que han estudiado diferentes cantidades de horas, y estas son sus calificaciones:

Horas de estudio (X)    Calificación (Y)
1                       65
2                       68   
3                       72
4                       75
5                       78
6                       82   
7                       86
8                       89

La regresión lineal simple busca establecer una relación lineal entre una variable independiente (X) y una variable dependiente (Y), representada por la ecuación:
Y = β₀ + β₁X + ε

Y es la variable dependiente (calificación)
X es la variable independiente (horas de estudio)
β₀ es el intercepto (valor esperado de Y cuando X = 0)
β₁ es la pendiente (cambio en Y por cada unidad de X)
ε es el error aleatorio

Para encontrar los mejores valores de β₀ y β₁, se utiliza el método de mínimos cuadrados, que minimiza la suma de los cuadrados de las diferencias entre los valores observados y los valores predichos:
Minimizar: Σ(Yᵢ - (β₀ + β₁Xᵢ))²
Esto nos lleva a las siguientes fórmulas:
β₁ = Σ((Xᵢ - X̄)(Yᵢ - Ȳ)) / Σ((Xᵢ - X̄)²)
β₀ = Ȳ - β₁X̄
Donde X̄ y Ȳ son las medias de X e Y respectivamente.

En nuestro ejemplo, al aplicar estas fórmulas (que el código hace automáticamente a través de scikit-learn), obtendremos valores aproximados de:

β₀ ≈ 61.5 (intercepto)
β₁ ≈ 3.5 (pendiente)

Esto significa que:

Un estudiante que no estudia (0 horas) tendría una calificación base de aproximadamente 61.5 puntos
Por cada hora adicional de estudio, la calificación aumenta en aproximadamente 3.5 puntos

Error cuadrático medio (MSE): Mide la media de los cuadrados de los errores entre los valores observados y predichos.

Un valor más pequeño indica un mejor ajuste.

Coeficiente de determinación (R²): Indica la proporción de la varianza en la variable dependiente que es predecible a partir de la variable independiente.

Varía entre 0 y 1, donde valores más cercanos a 1 indican un mejor ajuste.
En nuestro ejemplo, R² debería ser alto (alrededor de 0.99), indicando una fuerte relación lineal.

Aplicacion practica:
Con esta ecuación Y = 61.5 + 3.5X, podemos:

Predecir la calificación para cualquier cantidad de horas de estudio
Entender cuántas horas adicionales necesitaría estudiar un alumno para mejorar su calificación en cierta cantidad de puntos
Establecer expectativas realistas sobre el rendimiento basado en el tiempo invertido

Este modelo simple ilustra perfectamente el concepto de regresión lineal y cómo puede utilizarse para cuantificar relaciones entre variables y hacer predicciones basadas en datos.
